proposals/kapp-controller/004-dependency-resolution/README.md

```diff
@@ -0,0 +1,619 @@
+        name: pkg-demo.microservice-v2
+        namespace: default
+      conditions:
+      - type: OwnedPackageInstall
+
+    - name: optional-dependency
+      wantedPackageRef:
+        refName: pkg6.test.carvel.dev
+        versionSelection:
+          constraints: 0.6.0
+      conditions:
+      - type: UnknownPackage
+
+    - name: dep-with-wrong-version
+      wantedPackageRef:
+        refName: pkg1.test.carvel.dev
+        versionSelection:
+          constraints: "1.0.1" # one on the cluster is 0.1.1
+      packageInstallRef:
+        name: pkg-demo.pkg1.test.carvel.dev
+      conditions:
+      - type: FoundPackageInstall
+      - type: FoundPackageInstallRequiresDifferentVersion
+
+    - name: dep-with-unknown-package
+      wantedPackageRef:
+        refName: pkg1.test.carvel.dev
+      conditions:
+      - type: UnknownPackage
+
+    - name: dep-with-package-with-high-version
+      wantedPackageRef:
+        refName: pkg1.test.carvel.dev
+        versionSelection:
+          constraints: ">4.0.0" # cant calculate exact version in this system
+      conditions:
+      - type: UnknownPackageVersion
+
+```
+
+Example of a PackageInstall that is automatically created
+```yaml
+apiVersion: packaging.carvel.dev/v1alpha1
+kind: PackageInstall
+metadata:
+  name: pkg1.test.carvel.dev.1.0.1
+  annotations:
+    kapp-controller.carvel.dev/owner: "PackageInstall/pkg-demo"
+spec:
+    serviceAccountName: default-ns-sa
+  packageRef:
+    refName: pkg3.test.carvel.dev
+    versionSelection:
+      constraints: 1.0.0
+  values:
+    secretRef:
+      name: pkg3.dev-values
+...
+```
+
+#### Dependency resolution
+The dependency resolution will be done one PackageInstall at a time, basically each PackageInstall will only create the PackageInstalls that it directly depends on.
+
+ex:
+```
+When pkg depends on pkg1 and pkg3, while pkg1 depends on pkg2
+kapp-controller when it is processing pkg it will only
+create PackageInstalls for pkg1 and pkg3.
+
+pkg
+  -> pkg1
+      -> pkg2
+  -> pkg3
+```
+
+For future versions we can contemplate the possibility of each PackageInstall checking if the dependencies it has can be fulfilled with the Packages present in the Cluster.
+
+#### Use Cases
+
+##### Package depends on another Package
+- When the Package that match the dependency is already installed, no new installation is needed
+- When the dependent Pacakage is not installed, it will try to install it
+- When a dependent Package is already installed but the version does not fullfil the requirements of the current PackageInstall, kapp-controller will try to install the version that matches
+- When there is no Package that fullfills the dependency of a PackageInstall the installation will fail
+
+##### Two Packages depend on the same Package
+In this scenario the first PackageInstall that creates the dependent PackageInstall will own it. This means that the second PackageInstall will not need to create the dependent PackageInstall
```

Comment:
How does this work with transitive dependencies? Pkg A -> pkg B - > pkg C (with `->` signifying one depends on another)? I assume pkg C would be owned by pkg B (but it would be nice to add some info on the same).

Reply:
Sure, I will add more info around it. The current plan is to only deal with first level of dependency in the first iteration and then think about transitive dependencies in further iterations. 
However, even with one level of dependency, since by default kapp-controller will try to install the dependencies, even pkg C would be installed, and yes, it would be owned by pkg B (will add details around it).